tensorflow/compiler/xla/service/gpu/gemm_rewriter.cc

```diff
@@ -629,26 +653,103 @@ class GemmRewriterVisitor : public DfsHloRewriteVisitor {
+                    .lhs_batch_dimensions()
+                    .size() !=
+            2 ||
+        (b_bitcast ? b_bitcast : b)->shape().dimensions_size() -
+                gemm_backend_config.dot_dimension_numbers()
+                    .rhs_batch_dimensions()
+                    .size() !=
+            2) {
+      return false;
+    }
+
+    // Shift any bitcasts to the unconverted and unscaled operands.
+    if (a_bitcast) {
+      a = instr->AddInstruction(a_bitcast->CloneWithNewOperands(
+          ShapeUtil::MakeShape(a->shape().element_type(),
+                               a_bitcast->shape().dimensions()),
+          {a}));
+    }
+    if (b_bitcast) {
+      b = instr->AddInstruction(b_bitcast->CloneWithNewOperands(
+          ShapeUtil::MakeShape(b->shape().element_type(),
+                               b_bitcast->shape().dimensions()),
+          {b}));
+    }
+
     // cuBLASLt FP8 GEMM kernels currently require the first operand, i.e. A, to
     // be transposed. If the result of the GEMM is not in column major order, A
     // and B are later exchanged, and B is transposed here instead.
     // TODO(philipphack): Remove once cuBLASLt supports the NN configuration.
-    TF_ASSIGN_OR_RETURN(auto gemm_config,
-                        instr->backend_config<GemmBackendConfig>());
     TF_ASSIGN_OR_RETURN(bool is_col_major,
-                        OutputIsColumnMajor(instr, gemm_config));
+                        OutputIsColumnMajor(instr, gemm_backend_config));
+
+    // Identify the dimensional order which describes a transpose of the
+    // contracting and non-contracting dimensions of the GEMM.
+    auto transp_dim_order =
+        [](HloInstruction *x, absl::Span<const int64_t> x_contracting_dims,
+           absl::Span<const int64_t> x_batch_dims) -> std::vector<int64_t> {
+      std::vector<int64_t> dims;
+      dims.reserve(x->shape().dimensions_size());
+      for (int i = 0; i < x->shape().dimensions_size(); ++i) {
+        dims.emplace_back(-1);
+      }
+      // Discard the batch dimensions.
+      for (int64_t batch_dim : x_batch_dims) {
+        dims[batch_dim] = batch_dim;
+      }
+      // Identify the non-contracting dimension.
+      int non_contracting_dim;
+      for (int i = 0; i < x->shape().dimensions_size(); ++i) {
+        if (dims[i] == -1 && !absl::c_linear_search(x_contracting_dims, i)) {
+          non_contracting_dim = i;
+        }
+      }
+      // Exchange the contracting and non-contracting dimensions.
+      dims[non_contracting_dim] = x_contracting_dims[0];
+      dims[x_contracting_dims[0]] = non_contracting_dim;
+      return dims;
+    };
+
+    auto transp_dims =
+        [](HloInstruction *x,
+           absl::Span<const int64_t> transp_dim_order) -> std::vector<int64_t> {
+      std::vector<int64_t> transp_dims;
+      transp_dims.reserve(x->shape().dimensions_size());
+      for (int64_t dim : transp_dim_order) {
+        transp_dims.emplace_back(x->shape().dimensions(dim));
+      }
```

Comment:
This actually computes the dims for the inverse of the transpose. This is functionally correct because you only swap two dimensions in practice, but is confusing and should be fixed.

This can be fixed by replacing this lambda function with [`ShapeUtil::PermuteDimensions`](https://cs.opensource.google/tensorflow/tensorflow/+/master:tensorflow/compiler/xla/shape_util.h;l=541;drc=fd9f2505fb3cedbf17dd9a69f67b0935f6c27b46)

Reply:
I think PermuteDimensions implements the transpose through bitcasting and changing the memory layout, which is probably not what we want here because we have to actually transpose the matrix in memory.

What we currently have is essentially
`transp_dims[i] = x->shape().dimensions(transp_dim_order[i]);`

Wouldn’t the inverse of the transpose be something like
`transp_dims[transp_dim_order[i]] = x->shape().dimensions(i);
`

Reply:
You are right. I'm not sure why I thought otherwise before (transposing is surprisingly confusing).